### ISSUE TITLE
Function reference in config.exs causes malformed sys.config

### ISSUE THREAD

**Issue by jannekai** [bitwalker/exrm#230]
When trying to generate a release for my project with mix release, I get the following error:

==> Generating release...
Cannot add file sys.config to tar file - [{error,{41,erl_parse,["syntax error before: ",["Fun"]]}}]

Copying the contents of rel/files/sys.config and evaluating file using erlang shell, it also failed with error:

* 41: syntax error before: Fun

Looking closer it seems like a function reference defined in config.exs is not being converted into a valid erlang function reference. In my case I am specifying a function pointer which overrides the default retry function with a custom one for erlcloud library.

In short, the config.exs has a section like:

```
	erlcloud_overrides: [
	  http_client: :hackney,
	  ddb_retry: &Aws.Ddb.retry/1,
	  ddb_initial_timeout: 5000
	]
```

Which gets convert in sys.config into:

```
   {erlcloud_overrides,
       [{http_client,hackney},
        {ddb_retry,#Fun<Elixir.Aws.Ddb.retry.1>},
        {ddb_initial_timeout,5000}]}]},
```

The correct format should probably be something like this (copied from erlcloud defaults):

```
ddb_retry=fun erlcloud_ddb_impl:retry/1::erlcloud_ddb_impl:retry_fun(),
```

Copied example from: https://github.com/gleber/erlcloud/blob/master/include/erlcloud_aws.hrl

**Comment by bitwalker** [comment]
The issue here is that `config.exs` is not parsed/read by exrm, but by `Mix.Config`. The function is being captured during evaluation, and then being written by exrm to the final `sys.config`. The only way to fix this would be to write my own parser for `config.exs`, which I do not want to do. My suggestion (if possible) would be to instead configure your app with an `{module, function, args}` tuple, and consume it via `apply/3`.

If we need a fix for this, I think it would have to be in Elixir as an option for `Mix.Config.read`, but I'm not sure what the approach would be. A hack on exrm's end that I can think of would be parsing `inspect fun` for the module/function/arity and replacing the value in the generated sys.config with what you described from `erlcloud`, but I'm not really happy with that as a solution.

/cc @josevalim

**Comment by josevalim** [comment]
@bitwalker is `fun Mod:Fun/Arity` even supported in config files?

In any case, you don't need your own parser, only your own dumper (which may be a good idea anyway so you can provide reasonable error messages).  You can easily detect if an anonymous function refers to a mod+fun+arity and dump it accordingly and ignore all other anonymous functions.

**Comment by jannekai** [comment]
You are absolutely right, using the function reference in config file was a bad idea. Didn't really think about too clearly how the configuration file system works when implementing it and since it works without problems while running with mix, thought it was supported. 

In any case I ended up with using simple :no_retry atom in config file and adding a function, which creates the function reference during runtime, when I'm applying changes to default erlcloud aws_config record.

```
  defp ddb_retry_func(config, overrides) do
    case Keyword.get overrides, :ddb_retry, nil do
      :no_retry ->
        &Aws.Ddb.retry/1
      nil ->
        aws_config config, :ddb_retry
    end
  end
```

The suggestion about better error messages is a good one, since initially I ended up reading relx code trying to find why tar operation failed...

**Comment by philipgiuliani** [comment]
Just came into the same issue!
